File: packages/dvai-bridge-android-llama-core/android/src/main/java/co/deepvoiceai/bridge/llama/core/PluginState.kt
package co.deepvoiceai.bridge.llama.core

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Owns the running state of the llama core: the model bridge, the HTTP
 * server, and the model metadata. All access serialized through the Mutex
 * (Kotlin's actor-equivalent for this use case).
 *
 * Capacitor neutrality: this class takes plain `Map<String, Any?>` for
 * `start(opts:)` and returns the same shape from `start` / `statusInfo` so
 * the core compiles without a Capacitor dependency. The Capacitor wrapper
 * (Plugin.kt in dvai-bridge-capacitor-llama) translates JSObject ↔ Map at
 * the JS-bridge boundary.
 */
class PluginState {
    private val mutex = Mutex()
    private var server: HttpServer? = null
    private var bridge: LlamaCppBridge? = null
    private var modelId: String = ""
    private var isRunning: Boolean = false
    private var baseUrl: String? = null
    private var port: Int? = null

    suspend fun start(opts: Map<String, Any?>): Map<String, Any?> = mutex.withLock {
        if (isRunning) stopInternal()

        val modelPath = (opts["modelPath"] as? String)
            ?: throw IllegalArgumentException("modelPath is required for llama backend")
        if (modelPath.isEmpty()) throw IllegalArgumentException("modelPath is required for llama backend")

        val mmprojPath = opts["mmprojPath"] as? String
        val chatTemplate = opts["chatTemplate"] as? String
        val gpuLayers = (opts["gpuLayers"] as? Number)?.toInt() ?: 99
        val contextSize = (opts["contextSize"] as? Number)?.toInt() ?: 2048
        val threads = (opts["threads"] as? Number)?.toInt() ?: 4
        val embeddingMode = opts["embeddingMode"] as? Boolean ?: false
        val httpBasePort = (opts["httpBasePort"] as? Number)?.toInt() ?: 38883
        val httpMaxPortAttempts = (opts["httpMaxPortAttempts"] as? Number)?.toInt() ?: 16
        val corsConfig = parseCors(opts["corsOrigin"])

        // Load model via bridge.
        val newBridge = LlamaCppBridge()
        val ok = newBridge.loadModel(modelPath, mmprojPath, gpuLayers, contextSize, threads, embeddingMode)
        if (!ok) {
            throw IllegalStateException("Failed to load model at $modelPath")
        }

        // Phase 2A Pass 2: load mmproj if provided. Failure is fatal — the
        // caller asked for a multimodal model and we couldn't deliver.
        if (!mmprojPath.isNullOrEmpty()) {
            val mmprojOk = newBridge.loadMmproj(mmprojPath)
            if (!mmprojOk) {
                newBridge.unload()
                throw IllegalStateException("Failed to load mmproj at $mmprojPath")
            }
        }
        val mmprojLoaded = newBridge.isMmprojLoaded()
        val modelHasAudioEncoder = mmprojLoaded && newBridge.hasAudioEncoder()

        // Wire handlers + ctx. Phase 2A Pass 2: real flags mirrored from the
        // bridge state. embeddingMode mirrors start opts so /v1/embeddings
        // can short-circuit when off. chatTemplate is optional Jinja override;
        // null/empty falls through to the model's bundled chat template.
        val handlers = LlamaHandlers(
            bridge = newBridge,
            modelId = modelPath,
            mmprojLoaded = mmprojLoaded,
            modelHasAudioEncoder = modelHasAudioEncoder,
            embeddingMode = embeddingMode,
            chatTemplate = chatTemplate,
        )
        val ctx = HandlerContext(modelId = modelPath, backendName = "llama")

        // Bind server + install routes
        val newServer = HttpServer()
        val boundPort = newServer.startWithRoutes(
            basePort = httpBasePort,
            maxAttempts = httpMaxPortAttempts,
            host = "127.0.0.1",
            handlers = handlers,
            ctx = ctx,
            corsConfig = corsConfig,
        )

        this.bridge = newBridge
        this.server = newServer
        this.modelId = modelPath
        this.port = boundPort
        this.baseUrl = "http://127.0.0.1:$boundPort/v1"
        this.isRunning = true

        return@withLock mapOf(
            "baseUrl" to "http://127.0.0.1:$boundPort/v1",
            "port" to boundPort,
            "backend" to "llama",
            "modelId" to modelPath,
        )
    }

    suspend fun stop() = mutex.withLock { stopInternal() }

    private suspend fun stopInternal() {
        server?.stop()
        bridge?.unload()
        server = null
        bridge = null
        modelId = ""
        baseUrl = null
        port = null
        isRunning = false
    }

    fun statusInfo(): Map<String, Any?> = buildMap {
        put("running", isRunning)
        baseUrl?.let { put("baseUrl", it) }
        if (isRunning) put("backend", "llama")
    }

    private fun parseCors(raw: Any?): CorsConfig = when (raw) {
        is String -> if (raw == "*") CorsConfig.Wildcard else CorsConfig.Exact(raw)
        is List<*> -> CorsConfig.Allowlist(raw.filterIsInstance<String>())
        else -> CorsConfig.Wildcard
    }
}

File: packages/dvai-bridge-ios-llama-core/ios/Sources/DVAILlamaCore/PluginState.swift
// Internal/PluginState.swift
import Foundation
#if !COCOAPODS
import DVAILlamaCoreObjC
#endif
#if !COCOAPODS
import DVAISharedCore
#endif

/// Owns the running state of the capacitor-llama plugin: the model bridge,
/// the HTTP server, and the model metadata. All access is serialised through
/// the actor isolation.
public actor PluginState {
    private var server: HttpServer?
    private var bridge: LlamaCppBridge?
    private(set) var modelId: String = ""
    private(set) var isRunning: Bool = false
    private(set) var baseUrl: String?
    private(set) var port: Int?

    public init() {}

    /// Start the plugin: load model, bind server, install routes.
    /// - Returns dictionary suitable for Capacitor's `call.resolve(...)`.
    public func start(opts: [String: Any]) async throws -> [String: Any] {
        if isRunning { try await stopInternal() }

        guard let modelPath = opts["modelPath"] as? String, !modelPath.isEmpty else {
            throw NSError(
                domain: "DVAIBridgeLlama",
                code: 400,
                userInfo: [NSLocalizedDescriptionKey: "modelPath is required for llama backend"]
            )
        }

        let mmprojPath = opts["mmprojPath"] as? String
        let chatTemplate = opts["chatTemplate"] as? String
        let gpuLayers = opts["gpuLayers"] as? Int ?? 99
        let contextSize = opts["contextSize"] as? Int ?? 2048
        let threads = opts["threads"] as? Int ?? 4
        let embeddingMode = opts["embeddingMode"] as? Bool ?? false
        let httpBasePort = opts["httpBasePort"] as? Int ?? 38883
        let httpMaxPortAttempts = opts["httpMaxPortAttempts"] as? Int ?? 16
        let corsRaw = opts["corsOrigin"]
        let corsConfig = parseCors(corsRaw)

        // Load model via the ObjC++ bridge (real llama.cpp under the hood).
        let bridge = LlamaCppBridge()
        try bridge.loadModel(
            atPath: modelPath,
            mmprojPath: mmprojPath,
            gpuLayers: Int32(gpuLayers),
            contextSize: Int32(contextSize),
            threads: Int32(threads),
            embeddingMode: embeddingMode
        )

        // Phase 2A Pass 2: load mmproj (if provided) so multimodal handlers
        // can light up. A failed mmproj load is fatal for this start() call —
        // the caller asked for a multimodal model and we couldn't deliver.
        if let mmprojPath = mmprojPath, !mmprojPath.isEmpty {
            do {
                try bridge.loadMmproj(atPath: mmprojPath)
            } catch {
                bridge.unload()
                throw error
            }
        }
        let mmprojLoaded = bridge.isMmprojLoaded
        // Audio encoder support implies mmproj is loaded AND mtmd reports
        // an audio encoder is present in the projector.
        let modelHasAudioEncoder = mmprojLoaded && bridge.hasAudioEncoder()

        // Build handlers + context first; Hummingbird requires routes
        // to be registered at Application construction time, so the
        // installRoutes → tryBind order is mandatory. Phase 2A Pass 2:
        // real flags mirrored from the bridge state. embeddingMode
        // comes straight from the start opts so /v1/embeddings can
        // short-circuit when off. chatTemplate is an optional
        // Jinja-compatible override; nil/empty falls through to the
        // model's bundled `tokenizer.chat_template`.
        let handlers = LlamaHandlers(
            bridge: bridge,
            modelId: modelPath,
            mmprojLoaded: mmprojLoaded,
            modelHasAudioEncoder: modelHasAudioEncoder,
            embeddingMode: embeddingMode,
            chatTemplate: chatTemplate
        )
        let ctx = HandlerContext(modelId: modelPath, backendName: "llama")
        let server = HttpServer()
        await server.installRoutes(handlers: handlers, ctx: ctx, corsConfig: corsConfig)

        // Bind server (with port-fallback). If bind fails, release the
        // bridge so the loaded llama context doesn't leak until next
        // start().
        let port: Int
        do {
            port = try await server.tryBind(
                basePort: httpBasePort,
                maxAttempts: httpMaxPortAttempts,
                host: "127.0.0.1"
            )
        } catch {
            bridge.unload()
            throw error
        }

        self.bridge = bridge
        self.server = server
        self.modelId = modelPath
        self.port = port
        self.baseUrl = "http://127.0.0.1:\(port)/v1"
        self.isRunning = true

        return [
            "baseUrl": self.baseUrl!,
            "port": port,
            "backend": "llama",
            "modelId": modelPath,
        ]
    }

    /// Stop the plugin: release model, stop server.
    public func stop() async throws {
        try await stopInternal()
    }

    private func stopInternal() async throws {
        await server?.stop()
        bridge?.unload()
        server = nil
        bridge = nil
        modelId = ""
        baseUrl = nil
        port = nil
        isRunning = false
    }

    /// Snapshot of the current running state, suitable for Capacitor `call.resolve(...)`.
    public func statusInfo() -> [String: Any] {
        var dict: [String: Any] = ["running": isRunning]
        if let baseUrl = baseUrl { dict["baseUrl"] = baseUrl }
        if isRunning { dict["backend"] = "llama" }
        return dict
    }

    /// Parse the CORS option from the start opts dict.
    private func parseCors(_ raw: Any?) -> CORSConfig {
        if let s = raw as? String {
            return s == "*" ? .wildcard : .exact(s)
        }
        if let arr = raw as? [String] {
            return .allowlist(arr)
        }
        return .wildcard
    }
}
